OG-52: Describe how to retrieve static grid information (#7)
MeteoSwiss/opendata-forecast-data: README.md

### 2.5 Accessing Static Grid Information: Height, Longitude, and Latitude

### 🚧  Temporary Notice Work in Progress

Besides the current forecasting files, each catalog contains two static files. They store permanent information about the height of the half levels (HHL) in the vertical grid and
the center point coordinates of each triangle (CLON/CLAT) on the horizontal grid. Note that the forecasting GRIB files contain no information on height, longitude and latitude. They have to be determined via the static files HHL and CLON/CLAT.

> #### 2.5.1 How to access the height of a grid point

Reviewer: Standardize the use of title case throughout the README and avoid using 'how to' in headings following the rest of the instructions
```suggestion
#### 2.5.1 Accessing the Height of a Grid Point
```

> In the static HHL file one can obtain the height of the half levels of the vertical grid in meters above see level. In order to point a value from the data file of a given parameter to a height in meters above sea level, follow the steps below.

Reviewer: ```suggestion
In the static HHL file, the heights of the half levels of the vertical grid are provided in meters above see level. In order to associate a value from a data file (for a given parameter) to a height in meters above sea level, follow the steps below:
```

> - Verify that the key `uuidOfHGrid` (Universally Unique Identifier) of the data file and the HHL file match.

Reviewer: ```suggestion
- Verify that the `uuidOfHGrid` (Universally Unique Identifier) key in the data file matches the one in the HHL file.
```

> - Obtain the value for the key `level` and check the key `typeOfLevel` by listing the GRIB messages. If `typeOfLevel` is

Reviewer: We could add a small reference to the commands we described at the end of our section to guide users how to for example list the GRIB messages
```suggestion
- Retrieve the value for the `level` key and inspect the `typeOfLevel` key by listing the GRIB messages:
```

>     - **generalVertical**, the level number of the key `level` matches the half level in the HHL file and returns the height in meters above sea level.

Reviewer: ```suggestion
    - **generalVertical**: The value of `level` corresponds directly to a half level in the HHL file and gives the height in meters above sea level.
```

>     - **generalVerticalLayer**, the level number of the key `level` corresponds to the full level. To retrieve the height in meters above sea level, the user has to take the average height of the two half levels above and below the full level.

Reviewer: ```suggestion
    - **generalVerticalLayer**: The `level` value corresponds to a full level. To obtain the height in meters above sea level, average the heights of the two surrounding half levels (above and below).
```

>     - any other type of level, it is specified in meters and is self-explanatory.

Reviewer: ```suggestion
    - **Other types of level**: These are usually specified directly in meters and are self-explanatory.
```

> ❗ **NOTE**: Values at the same position refer to the same grid point.

Reviewer: Is this needed? How wouldn't it be the case?

> #### 2.5.2 How to access the longitude and latitude of a grid point

Reviewer: Same comment as for the previous title
```suggestion
#### 2.5.2 Accessing the Longitude and Latitude of a Grid Point
```

> The CLON/CLAT file stores the longitude and latitude of the center points of each triangle on the horizontal grid. When retrieving a data file in a jupyter notebook the load function includes fetching the CLON/CLAT values. To retrieve CLON/CLAT without a python environment, follow the septs below.

Reviewer: ```suggestion
The CLON/CLAT file stores the longitude and latitude of the center points of each triangle in the horizontal grid. When loading data using `get_from_ogd` from meteodata-lab (see the [example notebook](https://github.com/MeteoSwiss/opendata-nwp-demos/blob/first_example_from_file/01_retrieve_process_precip.ipynb)), the CLON/CLAT values are automatically retrieved. Otherwise, to retrieve CLON/CLAT manually, follow the steps below:
```

> - Verify that the key `uuidOfHGrid` (Universally Unique Identifier) of the data file and the CLON/CLAT file match.

Reviewer: ```suggestion
- Ensure that the `uuidOfHGrid` (Universally Unique Identifier) key in the data file matches the one in the CLON/CLAT file.
```

> - To each value of the data set attach the longitude from the GRIB message with the `shortName` tlon and the latitude from the GRIB message with the `shortName` tlat. Values at the same position corresond to each other.

Reviewer: ```suggestion
- For each data value, attach the corresponding:
 - Longitude from the GRIB message with `shortName = tlon`
 - Latitude from the GRIB message with `shortName = tlat`
Values at the same position corresond to each other.
```

Reviewer: Same comment for "Values at the same position corresond to each other." as above.

Reviewer: What do you mean by attach? We should provide more infos on how to combine them

Reviewer 2: Completely agree with @ninaburg on this point. Its really hard to regrid the grib2 files if you don't have the exact commands. Open support request with cdo here: https://code.mpimet.mpg.de/boards/1/topics/16251

Optimally the weights for regridding are provided as they do for the DWD ICON models: https://opendata.dwd.de/weather/lib/cdo/

Thank you very much for the great initiative!

Reviewer: Thanks for your comment, @MrMorice. We understand that regridding GRIB2 files can be challenging without precise commands. For now, we provide an API for regridding (see [Regriddind to a Regular Grid Notebook Example](https://github.com/MeteoSwiss/opendata-nwp-demos/blob/main/01_retrieve_process_precip.ipynb), and step-by-step solutions will be added progressively, especially as more feedback comes in once we’re in production

Reviewer: Coud you also address that comment please @nnellen ? https://github.com/MeteoSwiss/opendata-forecast-data/pull/7#discussion_r2011789918

As discussed during the standup it is enough to give the explanations how to dowload the static files, you could just give the information that the info is stored in the shorname tlat/tlon and if they want to have one file instead of two they can concatenate two files: 

cat grib_file_1 grib_file_2 > output_grib_file

- `forecast:perturbed`: Boolean flag determining if the data is deterministic (`false`) or ensemble-based.
- `forecast:horizon`: Defines the lead time of the forecast in ISO 8601 duration format (`P0DT00H00M00S` for instant data).

#### 2.6.2 Downloading the Forecast Data
Upon a successful request, the response will contain a dictionary of metadata, including forecast file links under the `assets` key. Locate the `href` field containing the pre-signed URL.
Download the GRIB file using the following command:
```
wget -O <desired_filename> “<pre-signed URL>”
```
Once downloaded, proceed with decoding the GRIB file using the instructions in [Section 2.7.4 Decoding GRIB Files with ecCodes](#274-decoding-grib-files-with-eccodes).

> #### 2.6.3 Installing ecCodes and COSMO definitions

Reviewer: Shouldn't we move " reading a forecast file using ecCodes" at the end, after the retrieval step—so the process would first cover downloading both the forecast and static files together, and then move on to how to read them

Author: Good idea.

#### 2.7.1 Accessing Vertical Grid Parameters

In the static vertical file, the heights of the half levels of the vertical grid are provided in meters above see level. In order to associate a value from a data file (for a given parameter) to a height in meters above sea level, follow the steps below:

1. Submit a GET request specifying the collection you want to retrieve the static vertical files from (e.g., `ch.meteoschweiz.ogd-forecasting-icon-ch1` for ICON-CH1-EPS):
```
curl -X GET https://sys-data.int.bgdi.ch/api/stac/v1/collections/ch.meteoschweiz.ogd-forecasting-icon-ch1/assets
```
2. Locate under `assets` in `id: vertical_constants_icon-ch1-eps.grib2` the `href` field and copy the pre-signed URL.
3. Download the file with:
```
wget -O <desired_filename> “<pre-signed URL>”
```
4. Once the static GRIB file is downloaded, verify that the `uuidOfHGrid` (Universally Unique Identifier) key in the data file matches the one in the HHL file.
5. Retrieve the value for the `level` key and inspect the `typeOfLevel` key by listing the GRIB messages:

>     - **generalVertical**: The value of `level` corresponds directly to a half level in the HHL file and gives the height in meters above sea level.

Reviewer: Sorry, I initially missed this: the level key (ranging from 1–81) does not directly correspond to the height in meters above sea level. Instead, it identifies the vertical level index (i.e., each GRIB message), and the actual height in meters is provided by the h field for each grid point.

Here's a suggested correction to the description to avoid confusion:
```suggestion
    - **generalVertical**: The value of `level` corresponds to a half level in the HHL file. For each level (i.e., each GRIB message), the variable `h` provides the height in meters above sea level for every grid point.
```

I'd add this fix in another PR not to block that PR.

- `forecast:perturbed`: Boolean flag determining if the data is deterministic (`false`) or ensemble-based.
- `forecast:horizon`: Defines the lead time of the forecast in ISO 8601 duration format (`P0DT00H00M00S` for instant data).

#### 2.6.2 Downloading the Forecast Data
Upon a successful request, the response will contain a dictionary of metadata, including forecast file links under the `assets` key. Locate the `href` field containing the pre-signed URL.
Download the GRIB file using the following command:
```
wget -O <desired_filename> “<pre-signed URL>”
```
Once downloaded, proceed with decoding the GRIB file using the instructions in [Section 2.7.4 Decoding GRIB Files with ecCodes](#274-decoding-grib-files-with-eccodes).

### 2.7 Accessing Static Grid Information: Height, Longitude, and Latitude

> Besides the current forecasting files, each catalog contains two static files. They store permanent information about the height of the half levels (HHL) in the vertical grid and

Reviewer: ```suggestion
Besides the current forecast files, each catalog contains two static files. They store permanent information about the height of the half levels (HHL) in the vertical grid and
```

> the center point coordinates of each triangle (CLON/CLAT) on the horizontal grid.

Reviewer: I’d suggest removing all the references to 'CLON/CLAT', as the files aren’t named that way, nor do we use those shortnames to retrieve them — we use 'tlon/tlat' instead ([L. 254](https://github.com/MeteoSwiss/opendata-forecast-data/pull/7/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R254))
```suggestion
the center point coordinates of each triangle on the horizontal grid.
```

> ❗ **NOTE**: The forecasting GRIB files contain no information on height, longitude and latitude. They have to be determined via the static vertical (HHL) and horizontal (CLON/CLAT) grid parameters file.

#### 2.7.1 Accessing Vertical Grid Parameters

> In the static HHL file, the heights of the half levels of the vertical grid are provided in meters above see level. In order to associate a value from a data file (for a given parameter) to a height in meters above sea level, follow the steps below:

Reviewer: Same comment as for CLON/CLAT, the file is not named HHL and the shortName refering to the height in the file is called h. I know we've been naming a lot internally those "CLON/CLAT/HHL" files but the users may rather refer to the actual files names we give (e.g vertical_constants_icon-ch1/2-eps.grib2 ) and the name of the variable in the file (h)
```suggestion
In the static file, the heights of the half levels of the vertical grid are provided in meters above see level. In order to associate a value from a data file (for a given parameter) to a height in meters above sea level, follow the steps below:
```

> - Submit a GET request specifying which model's asset should be downloaded (eg. `ch.meteoschweiz.ogd-forecasting-icon-ch1` for ICON-CH1-EPS).

Reviewer: Reformulate the "which asset" by "which collection" and Numered list could be better in steps instructions. Same comment for the horizontal grid's section 
```suggestion
1. Submit a GET request specifying the collection you want to retrieve the static files from (e.g., `ch.meteoschweiz.ogd-forecasting-icon-ch1` for ICON-CH1-EPS):
```

the center point coordinates of each triangle (CLON/CLAT) on the horizontal grid.

> ❗ **NOTE**: The forecasting GRIB files contain no information on height, longitude and latitude. They have to be determined via the static vertical (HHL) and horizontal (CLON/CLAT) grid parameters file.

#### 2.7.1 Accessing Vertical Grid Parameters

In the static HHL file, the heights of the half levels of the vertical grid are provided in meters above see level. In order to associate a value from a data file (for a given parameter) to a height in meters above sea level, follow the steps below:

- Submit a GET request specifying which model's asset should be downloaded (eg. `ch.meteoschweiz.ogd-forecasting-icon-ch1` for ICON-CH1-EPS).
```
curl GET https://sys-data.int.bgdi.ch/api/stac/v1/collections/ch.meteoschweiz.ogd-forecasting-icon-ch1/assets
```
- Locate under `assets` in `id: vertical_constants_icon-ch1-eps.grib2` the `href` field and copy the pre-signed URL.
- Download the file with:
```
wget -O <desired_filename> “<pre-signed URL>”
```

> - Once the static GRIB file is downloaded, verify that the `uuidOfHGrid` (Universally Unique Identifier) key in the data file matches the on in the HHL file.

Reviewer: typo
```suggestion
- Once the static GRIB file is downloaded, verify that the `uuidOfHGrid` (Universally Unique Identifier) key in the data file matches the one in the HHL file.
```

- Retrieve the value for the `level` key and inspect the `typeOfLevel` key by listing the GRIB messages:
    - **generalVertical**: The value of `level` corresponds directly to a half level in the HHL file and gives the height in meters above sea level.
    - **generalVerticalLayer**: The `level` value corresponds to a full level. To obtain the height in meters above sea level, average the heights of the two surrounding half levels (above and below).
    - **Other types of level**: These are usually specified directly in meters and are self-explanatory.

> #### 2.7.2 Accessing Horizontal Grid Parameters

Reviewer: why having changed lat/lon to horizontal grid param? i think it's better to be more specific and it's the only horizonal coordinates we're providing but happy to hear about your motivation to change
```suggestion
#### 2.7.2 Accessing the Longitude and Latitude of a Grid Point
```

Author: the files in the catalog are called "horizontal grid parameters"

> The CLON/CLAT file stores the longitude and latitude of the center points of each triangle in the horizontal grid. To retrieve CLON/CLAT manually, follow the septs below.

Reviewer: I've removed 'manually' from my previous comment—it's already clear that the user is doing it themselves. The typo is still here for "steps"

```suggestion
The CLON/CLAT file stores the longitude and latitude of the center points of each triangle in the horizontal grid. To retrieve CLON/CLAT, follow the steps below:
```
